### PR TITLE
Adding code to properly split_clone a v3 table.

### PR DESCRIPTION
**Pull request by alainjobart**
We need a vindex of cost < 2 on a table (that is numeric or hash).
Then we use it to split the rows.

@sougou you can take a look if you want, but I'm gonna keep working on it to add a few more things there.

---
This change is [Reviewable](https://reviewable.io/reviews/youtube/vitess/1634)

### CODE REVIEW

**Comment by sougou** [comment]
Actually, we should just go for the primary vindex (colvindexes[0]), because it will most likely be the sharding key (lowest cost). It's also guaranteed to be unique, which is another requirement.

File: go/vt/worker/split_clone.go [hunk 1]
```diff
@@ -463,6 +465,35 @@ func (scw *SplitCloneWorker) copy(ctx context.Context) error {
...
+		if err := json.Unmarshal([]byte(kschema), &kformal); err != nil {
+			return fmt.Errorf("error unmarshalling vschema for keyspace %s: %v", scw.keyspace, err)
+		}
+
+		formal := &vindexes.VSchemaFormal{
+			Keyspaces: map[string]vindexes.KeyspaceFormal{
+				scw.keyspace: kformal,
+			},
+		}
+		vschema, err := vindexes.BuildVSchema(formal)
```
Comment (sougou) [inline go/vt/worker/split_clone.go]: I think you already noticed. This code is very similar to vindexes.Validate. We could look at sharing.

I see the rest of the stuff is WIP. I'll comment as things develop.

**Comment by alainjobart** [comment]
@sougou this is ready for review.
@michael-berlin PTAL at the last commit, that's the change we just talked about. It's much simpler. May require some more parameters in automation framework.

File: go/vt/worker/key_resolver.go [hunk 2]
```diff
@@ -79,4 +75,97 @@ func (r *v2Resolver) keyspaceID(row []sqltypes.Value) ([]byte, error) {
...
+	// the primary vindex is most likely the sharding key, and has to
+	// be unique.
+	if len(tableSchema.ColVindexes) == 0 {
+		return nil, fmt.Errorf("no vindex definition for table %v", td.Name)
+	}
+	colVindex := tableSchema.ColVindexes[0]
+	if colVindex.Vindex.Cost() > 1 {
+		return nil, fmt.Errorf("primary vindex cost is too high for table %v", td.Name)
+	}
+	unique, ok := colVindex.Vindex.(vindexes.Unique)
```
Comment (sougou) [inline go/vt/worker/key_resolver.go]: Nit: use vindexes.IsUnique instead.

Comment (sougou) [inline go/vt/worker/key_resolver.go]: Ahh. Nvm. Looks like you need the Unique interface var also.

File: go/vt/worker/key_resolver.go [hunk 3]
```diff
@@ -79,4 +75,97 @@ func (r *v2Resolver) keyspaceID(row []sqltypes.Value) ([]byte, error) {
...
+		shardingColumnIndex: columnIndex,
+		vindex:              unique,
+	}, nil
+}
+
+// keyspaceID implements the keyspaceIDResolver interface.
+func (r *v3Resolver) keyspaceID(row []sqltypes.Value) ([]byte, error) {
+	v := row[r.shardingColumnIndex]
+	ids := []interface{}{v}
+	ksids, err := r.vindex.Map(nil, ids)
```
Comment (sougou) [inline go/vt/worker/key_resolver.go]: There may be some future use cases where someone may use a lookup based sharding key. If so, we may have to supply a VCursor that's capable of fetching data from db. Not an immediate requirement.

**Comment by sougou** [comment]
LGTM

File: go/vt/worker/key_resolver.go (hunk 3 above)
Comment (alainjobart) [inline go/vt/worker/key_resolver.go]: Then performance would be really bad, having to do a lookup for each row. We would probably not implement it this way...

**Comment by michael-berlin** [comment]
LGTM

Some minor comments. Feel free to fix them forward if you like.

---

Reviewed 4 of 4 files at r1, 2 of 2 files at r2, 2 of 2 files at r3, 2 of 2 files at r4, 5 of 5 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[go/vt/worker/diff_utils.go, line 84 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1634#-KFBINqxC7ejquR4e7Td:-KFBINqySbsrs1Jp7r85:865020964) ([raw file](https://github.com/youtube/vitess/blob/8a78ec96b9502242f7f9ea72b1569af15cc60321/go/vt/worker/diff_utils.go#L84)):*
nit: Missing period at the end.

---

*[go/vt/worker/split_diff_cmd.go, line 66 \[r6\]](https://reviewable.io:443/reviews/youtube/vitess/1634#-KFBotJnKjcT3ePdRhSq:-KFBotJoMS3IWf7nIYRF:771923399) ([raw file](https://github.com/youtube/vitess/blob/6576771e1fc2e8b93df12fb7379b584e04952eab/go/vt/worker/split_diff_cmd.go#L66)):*
optional: You could add the default to worker/defaults.go and reference it here.

---

*[go/vt/worker/split_diff_cmd.go, line 172 \[r6\]](https://reviewable.io:443/reviews/youtube/vitess/1634#-KFBoyY_UWIduD7ztPfm:-KFBoyY_UWIduD7ztPfn:771923399) ([raw file](https://github.com/youtube/vitess/blob/6576771e1fc2e8b93df12fb7379b584e04952eab/go/vt/worker/split_diff_cmd.go#L172)):*
optional: You could add the default to worker/defaults.go and reference it here.

---

*[go/vt/worker/split_diff_test.go, line 287 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1634#-KFBoHHq0OW_Y4a1IWrv:-KFBoHHq0OW_Y4a1IWrw:2042198093) ([raw file](https://github.com/youtube/vitess/blob/8a78ec96b9502242f7f9ea72b1569af15cc60321/go/vt/worker/split_diff_test.go#L287)):*
nit: half

---

*[go/vt/worker/split_diff_test.go, line 288 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1634#-KFBoNJCIXakvoYoZ4VI:-KFBoNJCIXakvoYoZ4VJ:-1596711340) ([raw file](https://github.com/youtube/vitess/blob/8a78ec96b9502242f7f9ea72b1569af15cc60321/go/vt/worker/split_diff_test.go#L288)):*
nit: SQL

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1634#-:-KFBttTyYh1rc9dZXecC:-6412672)*



[Approved with PullApprove](https://pullapprove.com/youtube/vitess/pull-request/1634/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=youtube/vitess)

**Pull request merged**